L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+  // checks.
+  void set_certificate_verifier(
+      std::shared_ptr<CertificateVerifier> certificate_verifier);
+
+  // Sets the crl provider, see CrlProvider for more details.
+  void set_crl_provider(std::shared_ptr<CrlProvider> crl_provider);
+
+  // Sets the minimum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will use TLS v1.2.
+  // @param tls_version: The minimum TLS version.
+  void set_min_tls_version(grpc_tls_version tls_version);
+  // Sets the maximum TLS version that will be negotiated during the TLS
+  // handshake. If not set, the underlying SSL library will use TLS v1.3.
+  // @param tls_version: The maximum TLS version.
+  void set_max_tls_version(grpc_tls_version tls_version);
+
+  // Sets a custom chain builder implementation that replaces the default chain
+  // building from the underlying SSL library.
+  void set_custom_chain_builder(std::shared_ptr<CustomChainBuilder>
+  chain_builder);
+}
+
+// Contains configurable options on the server side.
+class TlsServerCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Server side is required to use a provider, because server always needs to
+  // use identity certs.
+  explicit TlsServerCredentialsOptions(
+      std::shared_ptr<CertificateProviderInterface> certificate_provider)
+      : TlsCredentialsOptions() {
+    set_certificate_provider(certificate_provider);
+  }
+
+
+  // Sets option to request the certificates from the client.
+  // The default is GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE.
+  void set_cert_request_type(
+      grpc_ssl_client_certificate_request_type cert_request_type);
+}
+
+// Contains configurable options on the client side.
+// Client side doesn't need to always use certificate provider. When the
+// certificate provider is not set, we will use the root certificates stored
+// in the system default locations, and assume client won't provide any
+// identity certificates(single side TLS).
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certs.
+  // The default is true.
+  void set_verify_server_certs(bool verify_server_certs);
+
+// Builds TLS Credentials given TLS options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const grpc::experimental::TlsServerCredentialsOptions& options);
+```
+
+Here is an example of how to use this API:
+```c++
+/* Create option and set basic security primitives. */
+TlsChannelCredentialsOptions options = TlsChannelCredentialsOptions();
+options.set_verify_server_cert(true);
+options.set_min_tls_version(TLS1_2);
+options.set_max_tls_version(TLS1_3);
+
+// The core credentials APIs are still C, so this is still a grpc_channel_credential
```

Comment:
I'm not sure I understand what this sentence is saying: isn't this gRFC describing what the state of the code will look like after all of the changes in this gRFC have been implemented? In that case, should we just omit any reference to grpc_channel_credential? Or did I misunderstand?

Reply:
Ah good catch - the c-core API will still be `grpc_channel_credential` because that exists outside of the TLS space here, so it won't be refactored. But this is describing the C++ layer, which is why the code itself is `std::shared_ptr<ChannelCredentials>`, I just didn't remove this comment when making that change.